rfcs/0018-transition-frontier-extensions.md

```diff
@@ -0,0 +1,101 @@
+
+### New Transition Frontier Micro-diffs
+
+Transition Frontier Diffs will now be represented as smaller, composable micro-diffs rather than monolithic diffs like before. The primary advantage of this is composability, however it also helps to unify diffs with persistent diffs, allowing us to more easily implement the incremental hash computation in transition frontier and removes a layer of translation between diff formats. It also more easily allows us to defer extra computation into the Transition Frontier Extensions themselves by keeping the individual diffs light. Below is an psuedo-code implementation of the new micro-diffs.
+
+```ocaml
+type 'mutant diff =
+  | Breadcrumb_added : Breadcrumb.t -> {added: Breadcrumb.t; parent: Breadcrumb.t} diff
+  | Root_transitioned : {new: State_hash.t; garbage: Breadcrumb.t list} -> {previous: Root.t; new: Root.t; garbage: Breadcrumb.t list} diff
+  | Best_tip_changed : State_hash.t -> {previous: Breadcrumb.t; new: Breadcrumb.t} diff
+```
+
+Using these micro diffs, the Transition Frontier will return a list of diffs for every mutation. For instance, it is possible for the Transition Frontier to just have one diff `[Breadcrumb_added ...]`, or it could have upwards of three diffs `[Breadcrumb_added ...; Root_transitioned ...; Best_tip_changed ...]`.
+
+NOTE: The New\_root diff is no longer necessary as Transition Frontier roots are now provided to Extensions at initialization.
+
+### In-Code Documentation and Notices
+
+As part of this RFC, we will add some documentation comments to the code related to Extensions describing the guidelines for adding new extensions and what the concerns of an extension should be. This will mostly just be regurgating information from the section in here laying out those guidelines.
+
+### Implementation Details
+
+In order to break a dependency cycle between the Transition Frontier and its Extensions, the Transition Frontier will be split up into a base implementation and a final wrapper that glues the base implementation together with the Extensions. The base Transition Frontier will include nearly all of the Transition Frontier data structure internals (Breadcrumbs, Nodes, queries), but it will not include the final full function for adding breadcrumbs to the tree. Instead, it will provide two key functions that are part of adding breadcrumbs with the following signature:
+
+```ocaml
+(* [Diff.E.t] is the existential wrapper for a ['a Diff.t] *)
+val calculate_diffs : t -> Breadcrumb.t -> Diff.E.t list
+val process_diff : type mutant. t -> mutant Diff.t -> mutant
+```
+
+Extensions can then be defined using this base module. Then the final module can glue everything together, as shown in this psuedo-code:
+
+```ocaml
+module Extension = struct
+  module type S = sig
```

Comment:
Just to avoid ambiguity, can you fully write out the signature of `S`? It seems like the only functions that `Ext` has is `handle_diff` and `broadcast`.

Is `Ext.broadcast` just writing the view into a pipe?